Documentation/proposals/202304-silences.md

```diff
@@ -0,0 +1,131 @@
+* Related Tickets:
+  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
+  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
+* Other docs:
+  * n/a
+
+This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences
+configurations in the Kubernetes way.
+
+# Why
+
+Prometheus-operator does not have provide a way to automate the management of Alertmanager silences and there is no current standard/best practice way to do that.
+Moreover, setting an automation to manage silences has been reported as cumbersome by multiple users.
+or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).
+
+* Users using CI/CD jobs to manage silences have reported that this is cumbersome, since they need to either expose
+  Alertmanager via an ingress or using a service account to connect to the cluster, and this is even more complicated
+  in a multi-cluster setup.
+  However, there's folks that uses a Kubernetes Job to adds Silences,
+  In both situation, a CI/CD pipeline is needed in order to automate the execution, this includes managing the secrets and/or deprecation of code
+  In the Kubernetes Job way, teams need to build and deploy a container and monitor its status, which is an additional effort.
+  CI/CD pipeline is not the best option in multi-cluster setups, teams need to manage k8s access, secrets, and configs.
+
+![CI/CD K8s Job Approach](../img/CICD-k8s-job.png "CI/CD K8s Job Approach")
+
+* Users that use a standalone operator that implements a Silence CRD are in a better situation,
+  for example users have reported that they are using GitOps to fully manage the life cycle of Silences Custom resources,
+  which in fact brings benefits to the whole team because having a history and revisions in Git as well as constancy to manage the silences in a multi-tenant environment.
+
+Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)),
+because that requires to manage an extra component +CRDs on top of the Prometheus operator.
+Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.
+
+Adding support for `Silence` CRD in the Prometheus-operator will provide more flexibility to users in terms of choosing the tool to deploy their silences and will free users from managing a standalone component within the stack.
+
+## Pitfalls of the current solution
+
+Using Alertmanager API Directly comes with drawbacks:
+
+* Teams have to build an automation to add silences in a centralized manner
+* There is no input validation, which can lead to an invalid silence configuration
```

Comment:
I might be wrong, but I think that Alertmanager's API denies POST requests with invalid input. (definition can be seen [here](https://github.com/prometheus/alertmanager/blob/9a8d1f976e12b325ec47b84987a78b7845738be6/api/v2/openapi.yaml#L327-L347))

I understand that even the Alertmanager CLI, `amtool`, has [input validation](https://github.com/prometheus/alertmanager/blob/9a8d1f976e12b325ec47b84987a78b7845738be6/cli/silence_add.go#L73-L86) when adding a silence.

Could you elaborate on what kind of input validation that is lacking today?

Reply:
But In CI, there's no input validation, that's the intention.

Reply:
Changed